lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb

```diff
@@ -0,0 +1,354 @@
+              id: enterpise_fee_ids_filter
+            )
+          end
+          unless enterprise_fee_owner_ids_filter.empty?
+            @enterprise_fee_filtered_ids = @enterprise_fee_filtered_ids.where(
+              enterprise_id: enterprise_fee_owner_ids_filter
+            )
+          end
+          @enterprise_fee_filtered_ids = @enterprise_fee_filtered_ids.pluck(:id)
+        end
+
+        def enterpise_fee_filters?
+          enterpise_fee_ids_filter + enterprise_fee_owner_ids_filter != []
+        end
+
+        def enterpise_fee_ids_filter
+          ransack_params["enterprise_fee_id_in"]&.filter(&:present?) || []
+        end
+
+        def enterprise_fee_owner_ids_filter
+          ransack_params["enterprise_fee_owner_id_in"]&.filter(&:present?) || []
+        end
+
+        def query_result
+          # The objective is to group the orders by
+          # [entreprise_rate, tax_rate, supplier_id, distributor_id and order_cycle_id]
+
+          # The order.all_adjustment describes
+          #   - the enterprise fees applied on the order
+          #   - the enterprise fees applied on the line items
+          #   - all the taxes including, the taxes applied on the enterprise fees
+          # The access to enterprise fee tax rates is done in two steps
+          #   1.we'll need to store two attributes for each order.all_adjustemnt.enterprise_fee
+          #     a. originator_id will refer to the enterprise fee instance
+          #     b. id of the adjustment
+          #   2. order.all_adjustemnt.tax.where(adjustment_id: id,"Spree::Adjustment")
+          #     - this will return the tax applied on the enterprise fees
+          orders = report_line_items.list.map(&:order).uniq
+          orders.flat_map(&join_enterprise_fee)
+            .flat_map(&join_tax_rate)
+            .flat_map(&join_supplier)
+            .group_by(&group_keys)
+            .each(&change_root_to_order)
+        end
+
+        def join_enterprise_fee
+          proc do |order|
+            query = order
+              .all_adjustments
+              .enterprise_fee
+
+            if enterpise_fee_filters?
+              query = query.where(originator_id: enterprise_fee_filtered_ids)
+            end
+            query.group('originator_id')
+              .pluck("originator_id", 'array_agg(id)')
+              .map do |enterprise_fee_id, enterprise_fee_adjustment_ids|
+              {
+                enterprise_fee_id: enterprise_fee_id,
+                enterprise_fee_adjustment_ids: enterprise_fee_adjustment_ids,
+                order: order
+              }
+            end
+          end
+        end
+
+        def join_tax_rate
+          proc do |item|
+            tax_rate_ids = item[:order].all_adjustments.tax.where(
+              adjustable_id: item[:enterprise_fee_adjustment_ids],
+              adjustable_type: "Spree::Adjustment"
+            ).pluck(:originator_id)
+
+            tax_rate_ids << nil if tax_rate_ids.empty?
```

Comment:
I don't really understand why we are doing this ? is that because `join_tax_rate` needs to return a non empty hash ?

Reply:
We are grouping orders by `[entreprise_rate, tax_rate, supplier_id, distributor_id and order_cycle_id]`

if the order doesn't match any `tax_rate`, we want to have  `nil` in the grouping key:  `[entreprise_rate, nil, supplier_id, distributor_id and order_cycle_id]`

Reply:
Yes, it's like a SQL left join between `LineItems` X `Tax rates`